lib/streamlit/elements/button.py

```diff
@@ -92,6 +98,89 @@ def button(
             kwargs=kwargs,
         )
 
+    def download_button(
+        self,
+        label: str,
+        data: DownloadButtonDataType,
+        file_name: Optional[str] = None,
+        mime: Optional[str] = None,
+        key: Optional[Key] = None,
+        help: Optional[str] = None,
+        on_click: Optional[WidgetCallback] = None,
+        args: Optional[WidgetArgs] = None,
+        kwargs: Optional[WidgetKwargs] = None,
+    ) -> bool:
+        """Display a download button widget.
+
+        Parameters
+        ----------
+        label : str
+            A short label explaining to what this button is for.
+        data : str or bytes or file
+            The contents of the file to be downloaded.
+        file_name: str
+            An optional string to use as the name of the file to be downloaded,
+            eg. 'my_file.csv'. If file name is not specified, then we provide
+            a generic name for the download.
+        mime : str or None
+            The MIME type of the data. If None, defaults to "text/plain" or
+            "application/octet-stream" depending on the data type.
+        key : str or int
+            An optional string or integer to use as the unique key for the widget.
+            If this is omitted, a key will be generated for the widget
+            based on its content. Multiple widgets of the same type may
+            not share the same key.
+        help : str
+            An optional tooltip that gets displayed when the button is
+            hovered over.
+        on_click : callable
+            An optional callback invoked when this button is clicked.
+        args : tuple
+            An optional tuple of args to pass to the callback.
+        kwargs : dict
+            An optional dict of kwargs to pass to the callback.
+
+        Returns
+        -------
+        bool
+            If the button was clicked on the last run of the app.
+        """
+
+        key = to_key(key)
+        check_session_state_rules(default_value=None, key=key, writes_allowed=False)
+        if is_in_form(self.dg):
+            raise StreamlitAPIException(
+                f"`st.download_button()` can't be used in an `st.form()`.{FORM_DOCS_INFO}"
+            )
+
+        download_button_proto = DownloadButtonProto()
+
+        download_button_proto.label = label
+        download_button_proto.default = False
+
+        marshall_file(
+            self.dg._get_delta_path_str(), data, download_button_proto, mime, file_name
+        )
+
+        if help is not None:
+            download_button_proto.help = dedent(help)
+
+        def deserialize_button(ui_value, widget_id=""):
```

Comment:
I see that this method is copied and pasted from the past button code. Is there a reason why this is declared here? Do we use this method in the download button API?

Reply:
This method is used in the download button`register_widget` call, and it is a way to receive value from the frontend back to the backend.